kafka/producer.go

```diff
@@ -17,4 +17,176 @@
 
 package kafka
 
-// TODO
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"go.uber.org/zap"
+
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/plugin/kzap"
+
+	"github.com/elastic/apm-data/model"
+	"github.com/elastic/apm-queue/queuecontext"
+)
+
+// Encoder encodes a model.APMEvent to a []byte
+type Encoder interface {
+	// Encode accepts a model.APMEvent and returns the encoded representation.
+	Encode(model.APMEvent) ([]byte, error)
+}
+
+// RecordMutator mutates the record associated with the model.APMEvent.
+// If the RecordMutator returns an error, it is considered fatal.
+type RecordMutator func(model.APMEvent, *kgo.Record) error
+
+// ProducerConfig holds configuration for publishing events to Kafka.
+type ProducerConfig struct {
+	// Broker holds the (host:port) address of the Kafka broker to which
+	// events should be published.
+	Broker string
+
+	// ClientID to use when connecting to Kafka. This is used for logging
+	// and client identification purposes.
+	ClientID string
+	// Version is the software version to use in the Kafka client. This is
+	// useful since it shows up in Kafka metrics and logs.
+	Version string
+
+	// Logger is used for logging producer errors.
+	Logger *zap.Logger
+
+	// Encoder holds an encoding.Encoder for encoding events.
+	Encoder Encoder
+
+	// Sync can be used to indicate whether production should be synchronous.
+	Sync bool
+
+	// Mutators holds the list of RecordMutator applied to all the records sent
+	// by the producer. If any errors are returned, the producer will not
+	// produce and return the error in ProcessBatch.
+	Mutators []RecordMutator
+}
+
+// Validate checks that cfg is valid, and returns an error otherwise.
+func (cfg ProducerConfig) Validate() error {
+	var err []error
+	if cfg.Broker == "" {
+		err = append(err, errors.New("broker cannot be empty"))
+	}
+	if cfg.Logger == nil {
+		err = append(err, errors.New("logger cannot be nil"))
+	}
+	if cfg.Encoder == nil {
+		err = append(err, errors.New("encoder cannot be nil"))
```

Comment:
I think we should prefix the messages with the package name to keep it consistent with the current producer.
```suggestion
		err = append(err, errors.New("kafka: broker cannot be empty"))
	}
	if cfg.Logger == nil {
		err = append(err, errors.New("kafka: logger cannot be nil"))
	}
	if cfg.Encoder == nil {
		err = append(err, errors.New("kafka: encoder cannot be nil"))
```

Reply:
Thanks! Somehow I missed that part :facepalm: